ledger/src/executor.rs

```diff
@@ -0,0 +1,421 @@
+/// This means multiple threads can have a handle to the replayer and results are sent back to the
+/// correct replayer handle each time.
+impl ReplayerHandle {
+    pub fn new(
+        request_sender: Sender<(Sender<ReplayResponse>, ReplayRequest)>,
+        response_sender: Sender<ReplayResponse>,
+        response_receiver: Receiver<ReplayResponse>,
+    ) -> ReplayerHandle {
+        ReplayerHandle {
+            request_sender,
+            response_sender,
+            response_receiver,
+        }
+    }
+
+    pub fn send(
+        &self,
+        request: ReplayRequest,
+    ) -> result::Result<(), SendError<(Sender<ReplayResponse>, ReplayRequest)>> {
+        self.request_sender
+            .send((self.response_sender.clone(), request))
+    }
+
+    pub fn recv_and_drain(&self) -> result::Result<Vec<ReplayResponse>, RecvError> {
+        let mut results = vec![self.response_receiver.recv()?];
+        results.extend(self.response_receiver.try_iter());
+        Ok(results)
+    }
+}
+
+impl Replayer {
+    pub fn new(num_threads: usize) -> Replayer {
+        let (request_sender, request_receiver) = unbounded();
+        let threads = Self::start_replay_threads(num_threads, request_receiver);
+        Replayer {
+            threads,
+            request_sender,
+        }
+    }
+
+    pub fn handle(&self) -> ReplayerHandle {
+        let (response_sender, response_receiver) = unbounded();
+        ReplayerHandle::new(
+            self.request_sender.clone(),
+            response_sender,
+            response_receiver,
+        )
+    }
+
+    pub fn start_replay_threads(
+        num_threads: usize,
+        request_receiver: Receiver<(Sender<ReplayResponse>, ReplayRequest)>,
+    ) -> Vec<JoinHandle<()>> {
+        (0..num_threads)
+            .map(|i| {
+                let request_receiver = request_receiver.clone();
+                Builder::new()
+                    .name(format!("solReplayer-{}", i))
+                    .spawn(move || loop {
+                        match request_receiver.recv() {
+                            Ok((
+                                response_sender,
+                                ReplayRequest {
+                                    tx_and_idx,
+                                    bank,
+                                    transaction_status_sender,
+                                    replay_vote_sender,
+                                    cost_capacity_meter,
+                                    tx_cost,
+                                    log_messages_bytes_limit,
+                                    entry_callback,
+                                    batch_idx,
+                                },
+                            )) => {
+                                let mut timings = ExecuteTimings::default();
+
+                                let txs = vec![tx_and_idx.0];
+                                let mut batch =
```

Comment:
expecting some comments on this :)

Reply:
yeah.. right now im assuming the scheduler algorithm is perfect, doesn't have any lock failures. it seems like that's okay, but its a pretty big change from what the system does now